### ISSUE TITLE
[MEDIUM] rcpp tokenizer (src/tokenizer.cpp) has no real LLaMA-3 regex pre-tokenizer — BPE merges across whitespace/punctuation, diverging from HF reference token IDs

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#92]
## Location
`src/tokenizer.cpp` — `pre_tokenize()` and the BPE merge loop, plus the file's own header comment

## Description

The file header claims the regex pre-tokenizer is not implemented and the whole input is encoded as a single chunk:

> // Regex pre-tokenizer is NOT implemented — we encode the whole input
> // as a single chunk. This matches HF reference tokenization on clean
> // ASCII text and diverges on some mixed-script / punctuation-heavy
> // inputs.

But the actual `pre_tokenize()` implementation only splits digit runs into 3-digit blocks — it does **not** split on whitespace, punctuation, or CJK the way the LLaMA-3 / tiktoken regex pre-tokenizer does:

```cpp
// src/tokenizer.cpp
static std::vector<std::string> pre_tokenize(const std::string& text) {
    ...
    while (i < n) {
        if (is_digit(text[i])) {
            // break digits into 3-digit chunks
            ...
        } else {
            // grab everything else up to the next digit, as ONE chunk
            size_t j = i;
            while (j < n && !is_digit(text[j])) ++j;
            chunks.emplace_back(text, i, j - i);
            i = j;
        }
    }
}
```

The comment on `pre_tokenize` claims "Whitespace handling stays with byte-level BPE, which already matches the reference on everything we've tested" — but byte-level BPE merges are governed purely by merge rank, and with merges spanning whitespace (e.g. `" the"`, `"\n
"`, `" ."`) the encoder will happily merge *across* whitespace/punctuation boundaries within a chunk that the reference pre-tokenizer would have kept separate. The reference regex (roughly `'\s?\p{L}+|\s?\p{N}+|...|\s+(?!\S)|\s+`) exists precisely to prevent that.

The net effect: for any input where a whitespace- or punctuation-spanning merge outranks the within-word merges the reference would have produced, this encoder emits a **different token-ID sequence** than `tiktoken` / HuggingFace for the same text. Those wrong IDs are then fed to the model → silent output corruption that looks like "the model is just bad" rather than "the tokenizer disagrees with training."

## Impact

- Token IDs produced by `rcpp_tokenizer_encode` disagree with the reference LLaMA-3 tokenizer on an unbounded set of inputs (anything with whitespace/punctuation where a cross-boundary merge wins). The author's "matches on clean ASCII" claim is a best-case subset, not a guarantee.
- Because this is the runtime tokenizer for the pure-C++ path, any model whose prompt isn't re-tokenized identically to how it was tokenized at training/eval time will see a different distribution → degraded or garbled output, with no error signalling.
- The pre-tokenizer comment and the header comment contradict each other (header says "single chunk"; code does split digits), so it's unclear which behaviour was intended.

## Suggested fix direction

- Port the actual LLaMA-3 pre-tokenization regex (the `'s?\p{L}+|\s?\p{N}{1,3}|...` pattern tiktoken uses). There are MIT-licensed reference implementations to crib from.
- Add a differential test against `tiktoken` / HF `transformers` over a large corpus and gate releases on it, so divergence is caught rather than asserted away.
- At minimum, reconcile the two comments so it's clear this is a known gap, not an accidental half-implementation.

## Related
Independent of existing issues. Compounds the output-quality problems in #56 — even once the engine computes real logits, a divergent tokenizer will produce wrong tokens.

**Comment by bong-water-water-bong** [comment]
Not fixed yet — flagging the reasoning so it's a deliberate decision, not an oversight.

This one is genuinely larger than a patch: correctly porting the LLaMA-3 pre-tokenization regex (`\s?\p{L}+|\s?\p{N}{1,3}|...|\s+(?!\S)|\s+` plus the contraction/special-case handling) is substantial, and unlike the other fixes I can't verify correctness by eye — tokenization divergence is invisible without a differential test against `tiktoken` / HF `transformers` over a real corpus. Shipping a partially-correct regex that *looks* complete risks being worse than the current documented gap (the header comment already warns it diverges).

The right sequence is: (1) add a differential-test harness that tokenizes a large corpus with both this encoder and a reference and diffs the IDs, wire it into CI; (2) then port the regex with the harness catching regressions. I didn't want to drop an untested regex on `main` in this pass.

Happy to do the harness + regex port as a follow-up if you want — it's self-contained but non-trivial (probably 1 medium PR). Leaving open until that lands.
